File: .github/scripts/classify-agent-authorship.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "json"
require "optparse"

options = {
  github_output: nil,
}

OptionParser.new do |parser|
  parser.on("--github-output PATH", "Append key=value outputs for GitHub Actions") do |path|
    options[:github_output] = path
  end
end.parse!

input = ARGF.read

messages = input.each_line.map do |line|
  next if line.strip.empty?

  parsed = JSON.parse(line)
  if parsed.is_a?(Hash)
    parsed.dig("commit", "message") || parsed["message"]
  end
end.compact

required_patterns = {
  "co_author" => /^Co-Authored-By:\s*Maestro <maestro@evalops\.dev>\s*$/i,
  "version" => /^Maestro-Version:\s*\S.+$/i,
  "prompt_id" => /^Maestro-Prompt-Id:\s*\S.+$/i,
  "approvals_id" => /^Maestro-Approvals-Id:\s*\S.+$/i,
}

marker_pattern = /
  ^Co-Authored-By:\s*Maestro\s+<maestro@evalops\.dev>\s*$ |
  ^Maestro-(?:Version|Prompt-Id|Approvals-Id):
/ix

agent_commits = 0
human_commits = 0
incomplete_commits = 0

messages.each do |message|
  has_marker = message.lines.any? { |line| line.match?(marker_pattern) }

  unless has_marker
    human_commits += 1
    next
  end

  agent_commits += 1
  missing_required = required_patterns.values.any? do |pattern|
    message.lines.none? { |line| line.match?(pattern) }
  end
  incomplete_commits += 1 if missing_required
end

label =
  if agent_commits.positive? && human_commits.positive?
    "mixed-authorship"
  elsif agent_commits.positive?
    "agent-authored"
  else
    "human-authored"
  end

outputs = {
  "label" => label,
  "total_commits" => messages.length,
  "agent_commits" => agent_commits,
  "human_commits" => human_commits,
  "incomplete_agent_commits" => incomplete_commits,
}

outputs.each { |key, value| puts "#{key}=#{value}" }

if options[:github_output]
  File.open(options[:github_output], "a") do |file|
    outputs.each { |key, value| file.puts("#{key}=#{value}") }
  end
end

File: .github/scripts/classify_agent_authorship.py
#!/usr/bin/env python3
"""Classify PR commit authorship from a JSONL stream of commits.

Port of classify-agent-authorship.rb. Ruby is not present on every runner
image this reusable workflow runs on, so the classifier uses python3, which
is proven to exist on all of them. Standard library only.
"""

import argparse
import json
import re
import sys

# Ruby \s / \S semantics are ASCII-only; keep re.ASCII so unicode whitespace
# does not change trailer matching relative to the Ruby original.
FLAGS = re.IGNORECASE | re.ASCII

REQUIRED_PATTERNS = {
    "co_author": re.compile(r"^Co-Authored-By:\s*Maestro\s+<maestro@evalops\.dev>\s*$", FLAGS),
    "version": re.compile(r"^Maestro-Version:\s*\S.*$", FLAGS),
    "prompt_id": re.compile(r"^Maestro-Prompt-Id:\s*\S.*$", FLAGS),
    "approvals_id": re.compile(r"^Maestro-Approvals-Id:\s*\S.*$", FLAGS),
}

MARKER_PATTERN = re.compile(
    r"^Co-Authored-By:\s*Maestro\s+<maestro@evalops\.dev>\s*$"
    r"|^Maestro-(?:Version|Prompt-Id|Approvals-Id):",
    FLAGS,
)


def read_input(paths):
    if paths:
        chunks = []
        for path in paths:
            with open(path, encoding="utf-8") as handle:
                chunks.append(handle.read())
        return "".join(chunks)
    return sys.stdin.read()


def extract_messages(text):
    messages = []
    for line in text.splitlines():
        if not line.strip():
            continue
        parsed = json.loads(line)
        if isinstance(parsed, dict):
            commit = parsed.get("commit")
            message = commit.get("message") if isinstance(commit, dict) else None
            if message is None:
                message = parsed.get("message")
            if message is not None:
                messages.append(message)
    return messages


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--github-output",
        metavar="PATH",
        help="Append key=value outputs for GitHub Actions",
    )
    parser.add_argument("files", nargs="*")
    args = parser.parse_args()

    messages = extract_messages(read_input(args.files))

    agent_commits = 0
    untrailered_commits = 0
    incomplete_commits = 0

    for message in messages:
        lines = message.splitlines(keepends=True)
        has_marker = any(MARKER_PATTERN.search(line) for line in lines)

        if not has_marker:
            untrailered_commits += 1
            continue

        agent_commits += 1
        missing_required = any(
            not any(pattern.search(line) for line in lines)
            for pattern in REQUIRED_PATTERNS.values()
        )
        if missing_required:
            incomplete_commits += 1

    if agent_commits > 0 and untrailered_commits > 0:
        label = "mixed-authorship"
    elif agent_commits > 0:
        label = "agent-authored"
    else:
        label = "agent-assisted"

    outputs = {
        "label": label,
        "total_commits": len(messages),
        "agent_commits": agent_commits,
        "untrailered_commits": untrailered_commits,
        "human_commits": untrailered_commits,
        "incomplete_agent_commits": incomplete_commits,
    }

    for key, value in outputs.items():
        print(f"{key}={value}")

    if args.github_output:
        with open(args.github_output, "a", encoding="utf-8") as handle:
            for key, value in outputs.items():
                handle.write(f"{key}={value}\n")


if __name__ == "__main__":
    main()
